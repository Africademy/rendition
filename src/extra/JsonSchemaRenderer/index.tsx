import * as React from 'react';
import { forEach, keys, pick, get, isEqual, isArray } from 'lodash';
import jsone from 'json-e';
import ajv from 'ajv';
import asRendition from '../../asRendition';
import {
	DefaultProps,
	RenditionSystemProps,
	Theme,
	Tooltip,
} from '../../common-types';
import { Flex } from '../../components/Flex';
import Alert from '../../components/Alert';
import widgets, { WidgetWrapperUiOptions, WidgetMeta } from './widgets';
import { Value, JSONSchema, UiSchema, Format } from './types';

interface WrapperProp extends Tooltip {
	children: React.ReactNode;
}

const Wrapper = ({ tooltip, children, ...props }: WrapperProp) => (
	<Flex
		tooltip={tooltip}
		minWidth={0}
		minHeight={0}
		alignItems="flex-start"
		mb={1}
		flexDirection="column"
		{...props}
	>
		{children}
	</Flex>
);

export const getValue = (value?: Value, uiSchema?: UiSchema) => {
	return get(uiSchema, 'ui:value', value);
};

export const getType = (value?: Value) => {
	if (value === undefined) {
		return 'default';
	}
	if (value === null) {
		return 'null';
	}
	return isArray(value) ? 'array' : typeof value;
};

export const getWidget = (value?: Value, uiSchemaWidget?: string) => {
	const typeWidgets = get(widgets, getType(value), widgets.default);
	return get(typeWidgets, uiSchemaWidget || 'default', widgets.default.default);
};

const processValue = ({
	value,
	uiSchema,
}: Pick<InternalJsonSchemaRendererProps, 'value' | 'uiSchema'>) => {
	return getValue(value, uiSchema);
};

class JsonSchemaRenderer extends React.PureComponent<
	ThemedJsonSchemaRendererProps,
	JsonSchemaRendererState
> {
	validator: ajv.Ajv;
	validateFn: ajv.ValidateFunction;

	constructor(props: ThemedJsonSchemaRendererProps) {
		super(props);
		this.state = {
			error: null,
			validationErrors: null,
		};
		this.validator = new ajv();

		forEach(props.extraFormats, ({ name, format }) => {
			this.validator.addFormat(name, format);
		});
	}

	componentDidMount() {
		if (this.props.validate && !this.props.nested) {
			this.validate(true);
		}
	}

	componentDidUpdate(prevProps: ThemedJsonSchemaRendererProps) {
		if (this.props.validate) {
			const schemaChanged = !isEqual(prevProps.schema, this.props.schema);
			if (
				!this.props.nested &&
				(!isEqual(prevProps.value, this.props.value) || schemaChanged)
			) {
				this.validate(schemaChanged);
			}
		}
	}

	componentDidCatch(error: any) {
		this.setState({ error });
	}

	validate(compile: boolean = false) {
		try {
			if (compile) {
				this.validateFn = this.validator.compile(this.props.schema);
			}
			const isValid = this.validateFn(this.props.value);
			if (!isValid) {
				console.error(
					`JsonSchemaRenderer: Validation errors:\n${this.validator.errorsText(
						this.validateFn.errors,
					)}`,
				);
				this.setState({
					validationErrors: this.validateFn.errors,
				});
			} else {
				this.setState({
					validationErrors: null,
				});
			}
		} catch (error) {
			this.setState({
				error,
			});
		}
	}

	render() {
		const { value, schema, uiSchema = {}, ...props } = this.props;
		const { error, validationErrors } = this.state;

		const wrapperProps = pick(
			get(uiSchema, 'ui:options', {}),
			...keys(WidgetWrapperUiOptions),
		);

		if (error) {
			return (
				<Wrapper {...props} {...wrapperProps}>
					<Alert
						my={2}
						plaintext
						danger
						tooltip={get(error, 'message', error.toString())}
					>
						Error processing JSON data
					</Alert>
				</Wrapper>
			);
		}

		if (uiSchema === null) {
			return null;
		}
		const processedUiSchema = jsone(uiSchema, { source: value });
		const processedValue = processValue({
			value,
			uiSchema: processedUiSchema,
		});
		if (processedValue === undefined || processedValue === null) {
			return null;
		}
		const Widget = getWidget(processedValue, processedUiSchema['ui:widget']);
		return (
			<Wrapper {...props} {...wrapperProps}>
				{validationErrors && (
					<Alert
						my={2}
						plaintext
						danger
						tooltip={this.validator.errorsText(validationErrors)}
					>
						Invalid data/schema
					</Alert>
				)}
				<WidgetMeta schema={schema} uiSchema={uiSchema} />
				<Widget
					value={processedValue}
					schema={schema}
					uiSchema={processedUiSchema}
				/>
			</Wrapper>
		);
	}
}

interface JsonSchemaRendererState {
	error: Error | null;
	validationErrors: ajv.ErrorObject[] | null | undefined;
}

interface ThemedJsonSchemaRendererProps
	extends InternalJsonSchemaRendererProps {
	theme: Theme;
}

interface InternalJsonSchemaRendererProps extends DefaultProps {
	nested?: boolean;
	validate?: boolean;
	value?: Value;
	schema: JSONSchema;
	uiSchema?: UiSchema;
	extraFormats?: Format[];
}

export type JsonSchemaRendererProps = InternalJsonSchemaRendererProps &
	RenditionSystemProps;

export default asRendition<React.FunctionComponent<JsonSchemaRendererProps>>(
	JsonSchemaRenderer,
);
