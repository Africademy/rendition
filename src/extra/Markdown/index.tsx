import * as React from 'react';
import Txt, { TxtProps } from '../../components/Txt';
import Heading from '../../components/Heading';
import Link from '../../components/Link';
import unified from 'unified';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import rehypeReact from 'rehype-react';
import sanitize from 'rehype-sanitize';
import raw from 'rehype-raw';
import prism from '@mapbox/rehype-prism';
import Divider from '../../components/Divider';
import styled, { withTheme } from 'styled-components';
import gh from 'hast-util-sanitize/lib/github.json';
import { Theme } from '../../common-types';
import 'prismjs/themes/prism.css';
export { gh as defaultSanitizerOptions };

const MarkdownWrapper = styled(Txt)`
	* {
		box-sizing: border-box;
	}

	& code {
		background-color: rgba(27, 31, 35, 0.05);
		padding: 0.2em 0px;
		margin: 0px;
		border-radius: 3px;
		line-height: 1.2;
	}

	& pre {
		background-color: rgba(27, 31, 35, 0.05);
	}

	& pre > code {
		background: none;
	}

	& code,
	kbd,
	pre {
		font-family: ${(props) => props.theme.monospace};
	}

	& kbd {
		display: inline-block;
		padding: 3px 5px;
		font-size: 11px;
		line-height: 10px;
		color: #444d56;
		vertical-align: middle;
		background-color: #fafbfc;
		border: solid 1px #d1d5da;
		border-bottom-color: #c6cbd1;
		border-radius: 3px;
		box-shadow: inset 0 -1px 0 #c6cbd1;
	}

	dl {
		padding: 0;
	}

	dl dt {
		padding: 0;
		margin-top: 16px;
		font-size: 1em;
		font-style: italic;
		font-weight: 600;
	}

	dl dd {
		padding: 0 16px;
		margin-bottom: 16px;
	}

	td,
	th {
		padding: 0;
	}

	table {
		border-spacing: 0;
		border-collapse: collapse;
		display: block;
		width: 100%;
		overflow: auto;
	}

	table th {
		font-weight: 600;
	}

	table th,
	table td {
		padding: 6px 13px;
		border: 1px solid #dfe2e5;
	}

	table tr {
		background-color: #fff;
		border-top: 1px solid #c6cbd1;
	}

	table tr:nth-child(2n) {
		background-color: #f6f8fa;
	}

	blockquote {
		margin: 0;
	}

	blockquote,
	dl,
	table {
		margin-top: 0;
		margin-bottom: 16px;
	}

	blockquote {
		padding: 0 1em;
		color: #6a737d;
		border-left: 0.25em solid #dfe2e5;
	}

	blockquote > :first-child {
		margin-top: 0;
	}

	blockquote > :last-child {
		margin-bottom: 0;
	}
`;

type components = {
	[element: string]: React.ReactElement;
};

export const getProcessor = (
	componentOverrides?: components,
	sanitizerOptions?: any,
) => {
	return unified()
		.use(remarkParse, { gfm: true })
		.use(remarkRehype, { allowDangerousHtml: true })
		.use(prism)
		.use(raw)
		.use(
			sanitize,
			sanitizerOptions ?? {
				...gh,
				attributes: {
					...gh.attributes,
					'*': [...gh.attributes['*'], 'class', 'className'],
				},
			},
		)
		.use(rehypeReact, {
			createElement: React.createElement,
			components: {
				p: (props: any) => <Txt.p {...props} />,
				a: (props: any) => <Link {...props} />,
				h1: (props: any) => <Heading.h1 {...props} />,
				h2: (props: any) => <Heading.h2 {...props} />,
				h3: (props: any) => <Heading.h3 {...props} />,
				h4: (props: any) => <Heading.h4 {...props} />,
				h5: (props: any) => <Heading.h5 {...props} />,
				h6: (props: any) => <Heading.h6 {...props} />,
				hr: (props: any) => <Divider {...props} />,
				...componentOverrides,
			},
		});
};

type MarkdownProps = TxtProps & {
	children: string;
	componentOverrides?: components;
	theme: Theme;
	sanitizerOptions?: any; // https://github.com/syntax-tree/hast-util-sanitize#schema
};

export const MarkdownBase = ({
	children,
	componentOverrides,
	sanitizerOptions,
	...props
}: MarkdownProps) => {
	return (
		<MarkdownWrapper {...props}>
			{
				(getProcessor(componentOverrides, sanitizerOptions).processSync(
					children,
				) as any).result // type any because vFile types doesn't contains result, even though it should.
			}
		</MarkdownWrapper>
	);
};

export const Markdown = withTheme(MarkdownBase);
