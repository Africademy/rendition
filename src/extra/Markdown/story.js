import * as React from 'react'
import marked from 'marked'
import { storiesOf } from '@storybook/react'
import defaults from 'lodash/defaults'
import cloneDeep from 'lodash/cloneDeep'
import withReadme from 'storybook-readme/with-readme'
import source, {
  customizationSamples
} from '../../stories/assets/markdownSample'
import { Box, Card, Table, Txt } from '../../'
import { Markdown, defaultSanitizerOptions } from './index'
import Readme from './README.md'

const generateTableData = () => {
  return customizationSamples.map(({ markdown, sanitizerOptions }) => {
    const customSanitizerOptions = defaults(
      cloneDeep(sanitizerOptions || {}),
      defaultSanitizerOptions
    )
    return {
      'Specific Sanitizer Options': (
        <pre>{JSON.stringify(sanitizerOptions, null, 2)}</pre>
      ),
      Original: (
        <Card small>
          <Markdown>{markdown}</Markdown>
        </Card>
      ),
      Customized: (
        <Card small>
          <Markdown sanitizerOptions={customSanitizerOptions}>
            {markdown}
          </Markdown>
        </Card>
      )
    }
  })
}

storiesOf('Extra/Markdown', module)
  .addDecorator(withReadme(Readme))
  .add('Standard', () => {
    return (
      <Box m={3}>
        <Markdown>{source}</Markdown>
      </Box>
    )
  })
  .add('Custom Sanitizer', () => {
    const cellAttributes = { style: { verticalAlign: 'top' } }

    const columns = [
      { field: 'Specific Sanitizer Options', cellAttributes },
      { field: 'Original', cellAttributes },
      { field: 'Customized', cellAttributes }
    ]
    return (
      <Box m={3}>
        <Table sortable={false} columns={columns} data={generateTableData()} />
      </Box>
    )
  })
  .add('Custom Renderer', () => {
    const customRenderer = renderer => {
      renderer.link = function (_href, _title, _text) {
        const link = marked.Renderer.prototype.link.apply(this, arguments)
        return link.replace(/google/g, 'rewrittenLink')
      }

      return renderer
    }

    const markdown = '[https://www.google.com/](https://www.google.com/)'
    return (
      <Box m={3}>
        <Txt mb={2}>
          Original markdown: <Txt.span monospace>{markdown}</Txt.span>
        </Txt>
        <Markdown customizeRenderer={customRenderer}>{markdown}</Markdown>
      </Box>
    )
  })
