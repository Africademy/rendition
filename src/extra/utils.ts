import marked from 'marked';
import sanitizeHtml from 'sanitize-html';
import identity from 'lodash/identity';

const defaultRenderer = new marked.Renderer();

defaultRenderer.link = function(_href, _title, _text) {
	const link = marked.Renderer.prototype.link.apply(this, arguments);
	return link.replace('<a', '<a target="_blank" rel="noopener noreferrer"');
};

const markedOptions = (renderer: marked.Renderer) => {
	return {
		// Enable github flavored markdown
		gfm: true,
		breaks: true,
		headerIds: false,
		// Input text is sanitized using `sanitize-html` prior to being transformed by
		// `marked`
		sanitize: false,
		renderer,
	};
};

export const defaultSanitizerOptions = {
	allowedTags: sanitizeHtml.defaults.allowedTags.concat([
		'article',
		'aside',
		'del',
		'div',
		'h1',
		'h2',
		'header',
		'img',
		'input',
		'span',
	]),
	allowedAttributes: {
		a: ['href', 'name', 'target', 'rel'],
		img: ['src', 'class', 'title', 'alt'],
		input: [
			{
				name: 'type',
				multiple: false,
				values: ['checkbox'],
			},
			'disabled',
			'checked',
		],
		span: ['class'],
		div: ['class'],
		aside: ['class'],
		article: ['class'],
		header: ['class'],
	},
};

export const parseMarkdown = (
	text: string = '',
	sanitizerOptions: sanitizeHtml.IOptions = defaultSanitizerOptions,
	customizeRenderer: (
		defaultRenderer: marked.Renderer,
	) => marked.Renderer = identity,
) => {
	const html = marked(text, markedOptions(customizeRenderer(defaultRenderer)));
	const clean = sanitizeHtml(html, sanitizerOptions as any);
	return clean;
};
