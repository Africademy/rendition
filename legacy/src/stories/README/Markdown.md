# Markdown

A simple component for rendering [GitHub flavored markdown](https://github.github.com/gfm/). This component
sanitizes input.
This component is not loaded by default as it relies on a markdown parsing package 
that you may not want to include in your application.
You can load this component using:

```
import { Markdown } from 'rendition/dist/extra/Markdown';
```

[View story source](https://github.com/balena-io-modules/rendition/blob/master/src/stories/Markdown.js)

## Props

| Name   | Type   | Default   | Required   | Description   |
| ------ | ------ | --------- | ---------- | ------------- |
| `children`  | `string` | - | âœ“ | The markdown source that should be rendered |

Any other properties supplied are spread to the root element ([`Txt`](#txt)).

## Inheritance

The properties of the [`Txt`](#txt) component are also available.