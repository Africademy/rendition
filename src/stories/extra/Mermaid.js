import * as React from 'react'
import { storiesOf } from '@storybook/react'
import withReadme from 'storybook-readme/with-readme'
import { Box } from '../../'
import { Mermaid } from '../../extra/Mermaid'
import * as Readme from '../README/Mermaid.md'

const source = `
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
`

storiesOf('Extra/Mermaid', module)
  .addDecorator(story => <Box p={3}>{story()}</Box>)
  .addDecorator(withReadme(Readme))
  .add('Standard', () => {
    return (
      <React.Fragment>
        <Mermaid value={source} />
      </React.Fragment>
    )
  })
