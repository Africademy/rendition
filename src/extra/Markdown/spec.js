/* globals expect, describe, it */
import { mount } from 'enzyme'
import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from '../../../dist'
import { Markdown } from '../../../dist/extra/Markdown'
import suite from './specsuite'

describe.only('Markdown component', () => {
  it('should match the stored snapshot', () => {
    const component = renderer.create(
      <Provider>
        <Markdown>Hello world</Markdown>
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  suite.forEach(testCase => {
    it(testCase.name, () => {
      const wrapper = mount(
        <Provider>
          <Markdown sanitizerOptions={testCase.sanitizerOptions}>
            {testCase.source}
          </Markdown>
        </Provider>
      )

      // Unwrap the provider and markdown wrapper to get the HTML inside
      const result = wrapper
        .childAt(0)
        .render()
        .children()
        .html()
        .trim()

      expect(result).toEqual(testCase.expected)
    })
  })
})
