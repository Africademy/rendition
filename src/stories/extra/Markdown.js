import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { withScreenshot } from 'storybook-chrome-screenshot'
import withReadme from 'storybook-readme/with-readme'
import { Box, Provider } from '../../'
import { Markdown } from '../../extra/Markdown'
import Readme from '../README/Markdown.md'

import image from '../assets/GitHub-Mark-120px-plus.png'

const source = `
# Markdown
A simple component for rendering *GitHub flavored markdown*.

## Headers

# This is an \`<h1>\` tag
## This is an \`<h2>\` tag
###### This is an \`<h6>\` tag

## Emphasis

*This text will be italic*
_This will also be italic_

**This text will be bold**
__This will also be bold__

_You **can** combine them_

## Lists

### Unordered

* Item 1
* Item 2
  * Item 2a
  * Item 2b

### Ordered

1. Item 1
1. Item 2
1. Item 3
   1. Item 3a
   1. Item 3b

## Images

![Example Image](${image})

## Links

http://github.com - automatic!
[GitHub](http://github.com)

## Blockquotes

As Ash Ketchum said:

> Do you always need a reason to help somebody?

## Inline code

I think you should use an
\`<addr>\` element here instead.

## Code blocks

\`\`\`
const foo = () => {
  return 'bar'
}
\`\`\`

## Task Lists

- [x] list syntax required (any unordered or ordered list supported)
- [x] this is a complete item
- [ ] this is an incomplete item

## Tables

First Header | Second Header
------------ | -------------
Content from cell 1 | Content from cell 2
Content in the first column | Content in the second column


## Strikethrough

~~Crossed out~~

Any word wrapped with two tildes (like \`~~this~~\`) will appear crossed out.

## Inline HTML

<strong>An image:</strong>
<br>
<img src="${image}" />
<br>
<em>A link:</em>
<a target="_blank" href="http://github.com/">http://github.com/</a>
`

storiesOf('Extra/Markdown', module)
  .addDecorator(withReadme(Readme))
  .addDecorator(withScreenshot())
  .add('Standard', () => {
    return (
      <Provider>
        <Box m={3}>
          <Markdown>{source}</Markdown>
        </Box>
      </Provider>
    )
  })
