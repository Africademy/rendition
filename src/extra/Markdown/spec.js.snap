// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown component should match the stored snapshot 1`] = `
.c0 {
  font-family: 'Nunito',Arial,sans-serif;
  font-size: 14px;
  line-height: 1.5;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  font-family: 'Nunito',Arial,sans-serif;
  font-size: 14px;
  color: #2A506F;
}

.c1 h1,
.c1 h2,
.c1 h3,
.c1 h4,
.c1 h5,
.c1 h6,
.c1 button {
  font-family: CircularStd,Arial,sans-serif;
}

.c2 {
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  font-size: inherit;
  word-wrap: break-word;
}

.c2 .pl-c {
  color: #6a737d;
}

.c2 .pl-c1,
.c2 .pl-s .pl-v {
  color: #005cc5;
}

.c2 .pl-e,
.c2 .pl-en {
  color: #6f42c1;
}

.c2 .pl-smi,
.c2 .pl-s .pl-s1 {
  color: #24292e;
}

.c2 .pl-ent {
  color: #22863a;
}

.c2 .pl-k {
  color: #d73a49;
}

.c2 .pl-s,
.c2 .pl-pds,
.c2 .pl-s .pl-pse .pl-s1,
.c2 .pl-sr,
.c2 .pl-sr .pl-cce,
.c2 .pl-sr .pl-sre,
.c2 .pl-sr .pl-sra {
  color: #032f62;
}

.c2 .pl-v,
.c2 .pl-smw {
  color: #e36209;
}

.c2 .pl-bu {
  color: #b31d28;
}

.c2 .pl-ii {
  color: #fafbfc;
  background-color: #b31d28;
}

.c2 .pl-c2 {
  color: #fafbfc;
  background-color: #d73a49;
}

.c2 .pl-c2::before {
  content: '^M';
}

.c2 .pl-sr .pl-cce {
  font-weight: bold;
  color: #22863a;
}

.c2 .pl-ml {
  color: #735c0f;
}

.c2 .pl-mh,
.c2 .pl-mh .pl-en,
.c2 .pl-ms {
  font-weight: bold;
  color: #005cc5;
}

.c2 .pl-mi {
  font-style: italic;
  color: #24292e;
}

.c2 .pl-mb {
  font-weight: bold;
  color: #24292e;
}

.c2 .pl-md {
  color: #b31d28;
  background-color: #ffeef0;
}

.c2 .pl-mi1 {
  color: #22863a;
  background-color: #f0fff4;
}

.c2 .pl-mc {
  color: #e36209;
  background-color: #ffebda;
}

.c2 .pl-mi2 {
  color: #f6f8fa;
  background-color: #005cc5;
}

.c2 .pl-mdr {
  font-weight: bold;
  color: #6f42c1;
}

.c2 .pl-ba {
  color: #586069;
}

.c2 .pl-sg {
  color: #959da5;
}

.c2 .pl-corl {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #032f62;
}

.c2 .octicon {
  display: inline-block;
  vertical-align: text-top;
  fill: currentColor;
}

.c2 a {
  background-color: transparent;
  -webkit-text-decoration-skip: objects;
}

.c2 a:active,
.c2 a:hover {
  outline-width: 0;
}

.c2 strong {
  font-weight: inherit;
}

.c2 strong {
  font-weight: bolder;
}

.c2 h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.c2 img {
  border-style: none;
}

.c2 svg:not(:root) {
  overflow: hidden;
}

.c2 code,
.c2 kbd,
.c2 pre {
  font-family: 'Ubuntu Mono','Courier New',monospace;
  font-size: 1em;
}

.c2 hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible;
}

.c2 input {
  font: inherit;
  margin: 0;
}

.c2 input {
  overflow: visible;
}

.c2 [type='checkbox'] {
  box-sizing: border-box;
  padding: 0;
}

.c2 * {
  box-sizing: border-box;
}

.c2 input {
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
}

.c2 a {
  color: #0366d6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2 a:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c2 strong {
  font-weight: 600;
}

.c2 hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #dfe2e5;
}

.c2 hr::before {
  display: table;
  content: '';
}

.c2 hr::after {
  display: table;
  clear: both;
  content: '';
}

.c2 table {
  border-spacing: 0;
  border-collapse: collapse;
}

.c2 td,
.c2 th {
  padding: 0;
}

.c2 h1,
.c2 h2,
.c2 h3,
.c2 h4,
.c2 h5,
.c2 h6 {
  margin-top: 0;
  margin-bottom: 0;
}

.c2 h1 {
  font-size: 32px;
  font-weight: 600;
}

.c2 h2 {
  font-size: 24px;
  font-weight: 600;
}

.c2 h3 {
  font-size: 20px;
  font-weight: 600;
}

.c2 h4 {
  font-size: 16px;
  font-weight: 600;
}

.c2 h5 {
  font-size: 14px;
  font-weight: 600;
}

.c2 h6 {
  font-size: 12px;
  font-weight: 600;
}

.c2 p {
  margin-top: 0;
  margin-bottom: 10px;
}

.c2 blockquote {
  margin: 0;
}

.c2 ul,
.c2 ol {
  padding-left: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.c2 ol ol,
.c2 ul ol {
  list-style-type: lower-roman;
}

.c2 ul ul ol,
.c2 ul ol ol,
.c2 ol ul ol,
.c2 ol ol ol {
  list-style-type: lower-alpha;
}

.c2 dd {
  margin-left: 0;
}

.c2 code {
  font-family: 'Ubuntu Mono','Courier New',monospace;
  font-size: 1em;
}

.c2 pre {
  margin-top: 0;
  margin-bottom: 0;
  font-size: 1em;
  font-family: 'Ubuntu Mono','Courier New',monospace'Ubuntu Mono','Courier New',monospace;
  line-height: 10px;
  color: #444d56;
  vertical-align: middle;
  background-color: #fafbfc;
  border: solid 1px #d1d5da;
  border-bottom-color: #c6cbd1;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #c6cbd1;
}

.c2:checked + .radio-label {
  position: relative;
  z-index: 1;
  border-color: #0366d6;
}

.c2 .task-list-item {
  list-style-type: none;
}

.c2 .task-list-item + .task-list-item {
  margin-top: 3px;
}

.c2 .task-list-item input {
  margin: 0 0.2em 0.25em -1.6em;
  vertical-align: middle;
}

.c2 hr {
  border-bottom-color: #eee;
}

<div
  className="c0 c1"
>
  <div
    className="c2"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>Hello world</p>
",
      }
    }
  />
</div>
`;
